psydac/api/fem.py

```diff
@@ -916,16 +1029,60 @@ def __init__(self, expr, kernel_expr, *args, **kwargs):
         if vector_space.parallel and vector_space.cart.num_threads>1:
             num_threads = vector_space.cart._num_threads
 
+        if vector_space.parallel and vector_space.cart.is_comm_null:
+            self._free_args = ()
+            self._func      = do_nothing
+            self._args      = ()
+            self._expr      = expr
+            self._comm      = domain_h.comm
+            return
+
+        if isinstance(target, Boundary):
+            ext        = target.ext
+            axis       = target.axis
+        else:
+            ext        = None
+            axis       = None
+
+        is_rational_mapping = False
+        if not( mapping is None ):
+            is_rational_mapping = isinstance( mapping, NurbsMapping )
+            kwargs['mapping_space'] = mapping.space
+
+        self._mapping                 = mapping
+        self._is_rational_mapping     = is_rational_mapping
+        kwargs['is_rational_mapping'] = is_rational_mapping
+        kwargs['discrete_space']      = self._space
+
+        kwargs['comm']        = None
+        if vector_space.parallel:
+            kwargs['comm'] = vector_space.cart.comm
+
+        space_quad_order = [qo - 1 for qo in get_quad_order(self._space)]
+        quad_order       = [qo + 1 for qo in kwargs.pop('quad_order', space_quad_order)]
+
+        # this doesn't work right now otherwise. TODO: fix this and remove this assertion
+        assert np.array_equal(quad_order, get_quad_order(self.space))
+
         kwargs['num_threads'] = num_threads
+
         BasicDiscrete.__init__(self, expr, kernel_expr,  quad_order=quad_order, **kwargs)
 
-        # ...
+        self._comm       = domain_h.comm
         grid             = QuadratureGrid( self.space,  axis=axis, ext=ext)
         self._grid       = grid
         self._test_basis = BasisValues( self.space, nderiv = self.max_nderiv, trial=True, grid=grid, ext=ext)
 
         self._args = self.construct_arguments()
```

Comment:
Please make sure that this function does not change `self`

Reply:
done in [ec10ae7](https://github.com/pyccel/psydac/pull/213/commits/ec10ae70e2106c6cc339a5aa2ae5ac81c351a7f6)